src/rdkafka_partition.c

```diff
@@ -609,26 +609,23 @@ void rd_kafka_toppar_desired_del (rd_kafka_toppar_t *rktp) {
  * Append message at tail of 'rktp' message queue.
  */
 void rd_kafka_toppar_enq_msg (rd_kafka_toppar_t *rktp, rd_kafka_msg_t *rkm) {
+	int wakeup_fd, queue_len;
+	rd_kafka_t *rk;
 
 	rd_kafka_toppar_lock(rktp);
+	wakeup_fd = rktp->rktp_msgq_wakeup_fd;
+	rk = rktp->rktp_rkt->rkt_rk;
 	rd_kafka_msgq_enq(&rktp->rktp_msgq, rkm);
+	queue_len = rd_kafka_msgq_len(&rktp->rktp_msgq);
+        rd_kafka_toppar_unlock(rktp);
 #ifndef _MSC_VER
-        if (rktp->rktp_msgq_wakeup_fd != -1 &&
-            rd_kafka_msgq_len(&rktp->rktp_msgq) == 1) {
+        if (wakeup_fd >= 0 && queue_len == 1) {
                 char one = 1;
                 int r;
-                r = rd_write(rktp->rktp_msgq_wakeup_fd, &one, sizeof(one));
-                if (r == -1)
-                        rd_kafka_log(rktp->rktp_rkt->rkt_rk, LOG_ERR, "PARTENQ",
```

Comment:
why was this removed?

Reply:
I replaced it with the assert because I can't see an error that can actually happen except EINTER maybe which maybe we should separately handle.  I don't have a problem with the logging in general, but in this case it makes it hard to drop the lock earlier.  Since the log message includes the topic name we somehow need to have access to that in the exceptional case.  If we drop the lock before write() we don't know that string will even still be around right? so the alternative I can see is copying the string every time while holding the lock and that seems really bad.  I guess we could drop the topic name / partition from the message, but that would mean still fetching rk while holding the lock.

Reply:
We can't make any safe assumptions on what error we get from a system call (since the code runs on a plethora of different operating systems and versions), so we must be very careful in not doing anything drastic in these rare unknown cases.
Thus, we absolute don't want to crash the application if a non-essential wakeup file descriptor is not writable :)

Generally, asserts should only be used to guard for internal state and should never be used for external input.

Regarding the log locking:
The log message is safe without locking, because:
 * since all relevant objects (rktp, rkt, rk) are refcounted and our caller must hold a reference these objects will not go away from under our feet
 * `rktp->rktp_rkt` and `rkt->rkt_rk` are constant, they are never changed during the lifetime of these objects
 * same with `rkt_topic`, `rktp_partition`, etc.

The only thing that may change is `rktp->rktp_msgq_wakeup_fd` which must be protected by a lock, but since you've already extracted the fd into a variable we can use that instead.
This also shows us a valid scenario where the write may fail:
 * thread 1 is in enq_msg() having extracted the fd and getting ready to write
 * thread 2 removes the io_event fd from the queue and closes the fd
In this case thread 1's write() will fail, but the program should continue to operate since on the next call the enq_msg() the fd will no longer be set.